Changes to enable per channel requant.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37829 Some TODO fixes.
* #37626 Perf optimization for conv and gemm kernels.
* #37624 Added per channel separate test cases for fc and deconv tests.
* #37623 Changes to enable per channel support on dynamic linear.
* #37622 Enabled per channel quantized static linear/conv
* #37621 Added per channel kernels for depthwise conv.
* **#37620 Changes to enable per channel requant.**
* #37619 Enable per channel zero point.
* #37618 Interface changes to enable per channel quant.

Summary:
Now channel wise quantization is supported for linear/conv. Depthwise convs are still pending.
Approach is same as with zero points. Pointer to requantization array is passed and it is looked up using output_channel_index.
All the kernels are appropriately modified except for the depthwise ones.
Tests are altered to generate per channel zero points and requant scales.
Unit tests are replicated for conv tests to exercise per_channel conv. This was not strictly needed since conv kernels were changed such that they did per channel anyway. When per channels is not needed zero point and requant scale were same across channels. 
However for depthwise convolutions we will be using different set of kernels to do per channel, which required us to introduce per_channel member to conv_param structure, to know which kernels to use  for depthwise conv. 
Ensuing modifications were to keep everything in sync for both regular conv and depthwise so that we dont have caveat when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

kimishpatel committed:
Changes to enable per channel requant.

Summary:
Now channel wise quantization is supported for linear/conv.
Depthwise conv are still pending.
Tests are altered to generate per channel zero points and requant
scales.
All the kernels are fixed appropritately.
Added per_channel member to conv_param structure.
And replicated conv tests to exercise per_channel conv.
This was not strictly needed since conv kernels were changed
such that they did per channel anyway. When per channels is not needed
zp and scale were same across channels. This was to minimize code
duplicaiton as the perf impact is estimated (to be measured though) to
be low.
However this is not likely the case for depthwise convs. Thus they will
have separate kernels, which required us to introduce per_channel member
to conv_param structure, to know which kernels to apply for depthwise.
Ensuing modifications were to keep everything in
sync for both regular conv and depthwise so that we dont have caveat
when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv.
Depthwise conv are still pending.
Tests are altered to generate per channel zero points and requant
scales.
All the kernels are fixed appropritately.
Added per_channel member to conv_param structure.
And replicated conv tests to exercise per_channel conv.
This was not strictly needed since conv kernels were changed
such that they did per channel anyway. When per channels is not needed
zp and scale were same across channels. This was to minimize code
duplicaiton as the perf impact is estimated (to be measured though) to
be low.
However this is not likely the case for depthwise convs. Thus they will
have separate kernels, which required us to introduce per_channel member
to conv_param structure, to know which kernels to apply for depthwise.
Ensuing modifications were to keep everything in
sync for both regular conv and depthwise so that we dont have caveat
when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv.
Depthwise conv are still pending.
Tests are altered to generate per channel zero points and requant
scales.
All the kernels are fixed appropritately.
Added per_channel member to conv_param structure.
And replicated conv tests to exercise per_channel conv.
This was not strictly needed since conv kernels were changed
such that they did per channel anyway. When per channels is not needed
zp and scale were same across channels. This was to minimize code
duplicaiton as the perf impact is estimated (to be measured though) to
be low.
However this is not likely the case for depthwise convs. Thus they will
have separate kernels, which required us to introduce per_channel member
to conv_param structure, to know which kernels to apply for depthwise.
Ensuing modifications were to keep everything in
sync for both regular conv and depthwise so that we dont have caveat
when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv.
Depthwise conv are still pending.
Tests are altered to generate per channel zero points and requant
scales.
All the kernels are fixed appropritately.
Added per_channel member to conv_param structure.
And replicated conv tests to exercise per_channel conv.
This was not strictly needed since conv kernels were changed
such that they did per channel anyway. When per channels is not needed
zp and scale were same across channels. This was to minimize code
duplicaiton as the perf impact is estimated (to be measured though) to
be low.
However this is not likely the case for depthwise convs. Thus they will
have separate kernels, which required us to introduce per_channel member
to conv_param structure, to know which kernels to apply for depthwise.
Ensuing modifications were to keep everything in
sync for both regular conv and depthwise so that we dont have caveat
when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv.
Depthwise conv are still pending.
Tests are altered to generate per channel zero points and requant
scales.
All the kernels are fixed appropritately.
Added per_channel member to conv_param structure.
And replicated conv tests to exercise per_channel conv.
This was not strictly needed since conv kernels were changed
such that they did per channel anyway. When per channels is not needed
zp and scale were same across channels. This was to minimize code
duplicaiton as the perf impact is estimated (to be measured though) to
be low.
However this is not likely the case for depthwise convs. Thus they will
have separate kernels, which required us to introduce per_channel member
to conv_param structure, to know which kernels to apply for depthwise.
Ensuing modifications were to keep everything in
sync for both regular conv and depthwise so that we dont have caveat
when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv.
Depthwise conv are still pending.
Tests are altered to generate per channel zero points and requant
scales.
All the kernels are fixed appropritately.
Added per_channel member to conv_param structure.
And replicated conv tests to exercise per_channel conv.
This was not strictly needed since conv kernels were changed
such that they did per channel anyway. When per channels is not needed
zp and scale were same across channels. This was to minimize code
duplicaiton as the perf impact is estimated (to be measured though) to
be low.
However this is not likely the case for depthwise convs. Thus they will
have separate kernels, which required us to introduce per_channel member
to conv_param structure, to know which kernels to apply for depthwise.
Ensuing modifications were to keep everything in
sync for both regular conv and depthwise so that we dont have caveat
when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv.
Depthwise conv are still pending.
Tests are altered to generate per channel zero points and requant
scales.
All the kernels are fixed appropritately.
Added per_channel member to conv_param structure.
And replicated conv tests to exercise per_channel conv.
This was not strictly needed since conv kernels were changed
such that they did per channel anyway. When per channels is not needed
zp and scale were same across channels. This was to minimize code
duplicaiton as the perf impact is estimated (to be measured though) to
be low.
However this is not likely the case for depthwise convs. Thus they will
have separate kernels, which required us to introduce per_channel member
to conv_param structure, to know which kernels to apply for depthwise.
Ensuing modifications were to keep everything in
sync for both regular conv and depthwise so that we dont have caveat
when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv.
Depthwise conv are still pending.
Tests are altered to generate per channel zero points and requant
scales.
All the kernels are fixed appropritately.
Added per_channel member to conv_param structure.
And replicated conv tests to exercise per_channel conv.
This was not strictly needed since conv kernels were changed
such that they did per channel anyway. When per channels is not needed
zp and scale were same across channels. This was to minimize code
duplicaiton as the perf impact is estimated (to be measured though) to
be low.
However this is not likely the case for depthwise convs. Thus they will
have separate kernels, which required us to introduce per_channel member
to conv_param structure, to know which kernels to apply for depthwise.
Ensuing modifications were to keep everything in
sync for both regular conv and depthwise so that we dont have caveat
when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv.
Depthwise conv are still pending.
Tests are altered to generate per channel zero points and requant
scales.
All the kernels are fixed appropritately.
Added per_channel member to conv_param structure.
And replicated conv tests to exercise per_channel conv.
This was not strictly needed since conv kernels were changed
such that they did per channel anyway. When per channels is not needed
zp and scale were same across channels. This was to minimize code
duplicaiton as the perf impact is estimated (to be measured though) to
be low.
However this is not likely the case for depthwise convs. Thus they will
have separate kernels, which required us to introduce per_channel member
to conv_param structure, to know which kernels to apply for depthwise.
Ensuing modifications were to keep everything in
sync for both regular conv and depthwise so that we dont have caveat
when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv. Depthwise convs are still pending.
Approach is same as with zero points. Pointer to requantization array is passed and it is looked up using output_channel_index.
All the kernels are appropriately modified except for the depthwise ones.
Tests are altered to generate per channel zero points and requant scales.
Unit tests are replicated for conv tests to exercise per_channel conv. This was not strictly needed since conv kernels were changed such that they did per channel anyway. When per channels is not needed zero point and requant scale were same across channels. 
However for depthwise convolutions we will be using different set of kernels to do per channel, which required us to introduce per_channel member to conv_param structure, to know which kernels to use  for depthwise conv. 
Ensuing modifications were to keep everything in sync for both regular conv and depthwise so that we dont have caveat when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv. Depthwise convs are still pending.
Approach is same as with zero points. Pointer to requantization array is passed and it is looked up using output_channel_index.
All the kernels are appropriately modified except for the depthwise ones.
Tests are altered to generate per channel zero points and requant scales.
Unit tests are replicated for conv tests to exercise per_channel conv. This was not strictly needed since conv kernels were changed such that they did per channel anyway. When per channels is not needed zero point and requant scale were same across channels. 
However for depthwise convolutions we will be using different set of kernels to do per channel, which required us to introduce per_channel member to conv_param structure, to know which kernels to use  for depthwise conv. 
Ensuing modifications were to keep everything in sync for both regular conv and depthwise so that we dont have caveat when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv. Depthwise convs are still pending.
Approach is same as with zero points. Pointer to requantization array is passed and it is looked up using output_channel_index.
All the kernels are appropriately modified except for the depthwise ones.
Tests are altered to generate per channel zero points and requant scales.
Unit tests are replicated for conv tests to exercise per_channel conv. This was not strictly needed since conv kernels were changed such that they did per channel anyway. When per channels is not needed zero point and requant scale were same across channels. 
However for depthwise convolutions we will be using different set of kernels to do per channel, which required us to introduce per_channel member to conv_param structure, to know which kernels to use  for depthwise conv. 
Ensuing modifications were to keep everything in sync for both regular conv and depthwise so that we dont have caveat when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv. Depthwise convs are still pending.
Approach is same as with zero points. Pointer to requantization array is passed and it is looked up using output_channel_index.
All the kernels are appropriately modified except for the depthwise ones.
Tests are altered to generate per channel zero points and requant scales.
Unit tests are replicated for conv tests to exercise per_channel conv. This was not strictly needed since conv kernels were changed such that they did per channel anyway. When per channels is not needed zero point and requant scale were same across channels. 
However for depthwise convolutions we will be using different set of kernels to do per channel, which required us to introduce per_channel member to conv_param structure, to know which kernels to use  for depthwise conv. 
Ensuing modifications were to keep everything in sync for both regular conv and depthwise so that we dont have caveat when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv. Depthwise convs are still pending.
Approach is same as with zero points. Pointer to requantization array is passed and it is looked up using output_channel_index.
All the kernels are appropriately modified except for the depthwise ones.
Tests are altered to generate per channel zero points and requant scales.
Unit tests are replicated for conv tests to exercise per_channel conv. This was not strictly needed since conv kernels were changed such that they did per channel anyway. When per channels is not needed zero point and requant scale were same across channels. 
However for depthwise convolutions we will be using different set of kernels to do per channel, which required us to introduce per_channel member to conv_param structure, to know which kernels to use  for depthwise conv. 
Ensuing modifications were to keep everything in sync for both regular conv and depthwise so that we dont have caveat when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

dreiss commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/pack.h:
```diff
@@ -92,9 +97,39 @@ static inline void pytorch_pack_q8gemm_wrq(
+      if (kzp != 0) {
+        // This part fills the packed wights with zero points for output channels
+        // when they are not divisble by nr blocking parameter.
+        // In that case
```
Missing the end of this comment?

dreiss reviewed (approved):
Looked at all but the microkernels and it looks good except for minor comments.

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8conv/8x8-aarch64-neon.S:
```diff
@@ -102,9 +107,13 @@ BEGIN_FUNCTION pytorch_q8conv_ukernel_8x8__aarch64_neon
+    // - v26 = requantization_scale channels 0-3
+    // - v31 = requantization_scale channels 4-7
+    LD1 {v26.4s}, [x10], 16
+    LD1 {v30.4s}, [x10]
```
This is a long shot but I am wondering if you can do this and remove `LSL x9, x9, 2` and `ADD x10, x10, x9` above, in case the value of x9 and x10 do not matter after this point:

```
LD1 {v26.4s}, [x10], x9, lsl 2
LD1 {v30.4s}, [x10, 16]
```

kimishpatel replied:
Does that form exist?. I checked but it does not seem to me that the variant of LD1 exist where shift can be folded in.

kimishpatel replied:
Yes so for LDR yes there is. Not for LD1. But I think this I can use in base + offset calculation done via add.

AshkanAliabadi reviewed (commented):
Got half way through, will continue tomorrow.

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv. Depthwise convs are still pending.
Approach is same as with zero points. Pointer to requantization array is passed and it is looked up using output_channel_index.
All the kernels are appropriately modified except for the depthwise ones.
Tests are altered to generate per channel zero points and requant scales.
Unit tests are replicated for conv tests to exercise per_channel conv. This was not strictly needed since conv kernels were changed such that they did per channel anyway. When per channels is not needed zero point and requant scale were same across channels. 
However for depthwise convolutions we will be using different set of kernels to do per channel, which required us to introduce per_channel member to conv_param structure, to know which kernels to use  for depthwise conv. 
Ensuing modifications were to keep everything in sync for both regular conv and depthwise so that we dont have caveat when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/fc-prepack.cc:
```diff
@@ -5,6 +5,61 @@
 #include <cstdlib>
 
 namespace qnnpack {
+// For runtime quantization packing.
```
Dynamic quantization is when you quantize your input data every single time and dequantize the output to fp32. Runtime quantization is the other mode. Only reason it is called runtime is due to the way QNNPACK was integrated in pytorch.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8conv/8x8-aarch64-neon.S:
```diff
@@ -57,8 +57,12 @@ BEGIN_FUNCTION pytorch_q8conv_ukernel_8x8__aarch64_neon
     # Load pointer to per channel zero points array
     LDR x10, [x8]
+    # To go to a_zero_point
+    ADD x8, x8, 8
```
Doesnt compile on android.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8conv/8x8-aarch64-neon.S:
```diff
@@ -67,10 +71,11 @@ BEGIN_FUNCTION pytorch_q8conv_ukernel_8x8__aarch64_neon
-    ADD x8, x8, 4
+    # Load pointer to per channel requant scale
+    LDR x10, [x8, 8]!
+    ADD x8, x8, 8
```
Same problem as before with LDRs.

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv. Depthwise convs are still pending.
Approach is same as with zero points. Pointer to requantization array is passed and it is looked up using output_channel_index.
All the kernels are appropriately modified except for the depthwise ones.
Tests are altered to generate per channel zero points and requant scales.
Unit tests are replicated for conv tests to exercise per_channel conv. This was not strictly needed since conv kernels were changed such that they did per channel anyway. When per channels is not needed zero point and requant scale were same across channels. 
However for depthwise convolutions we will be using different set of kernels to do per channel, which required us to introduce per_channel member to conv_param structure, to know which kernels to use  for depthwise conv. 
Ensuing modifications were to keep everything in sync for both regular conv and depthwise so that we dont have caveat when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/8x8-aarch64-neon.S:
```diff
@@ -99,9 +105,12 @@ BEGIN_FUNCTION pytorch_q8gemm_ukernel_8x8__aarch64_neon
+    // - v26 = requantization_scale channels 0-3
+    // - v27 = requantization_scale channels 4-7
+    LD1 {v26.4s}, [x17], 16
 
```
At a glance it seems to me that you can save a couple of instructions in the block above if interested.  It basically boils down to modifying the pointer arithmetic to take advantage of the free offsetting in the loads.

kimishpatel replied:
Yes so pointer arithmetic can save lsl by folding it in. LD1 I dont think since it has only no offset and post-index variants.

kimishpatel replied:
I am gonna make these changes in perf opt PR.

AshkanAliabadi reviewed (commented):
LGTM. thanks.

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv. Depthwise convs are still pending.
Approach is same as with zero points. Pointer to requantization array is passed and it is looked up using output_channel_index.
All the kernels are appropriately modified except for the depthwise ones.
Tests are altered to generate per channel zero points and requant scales.
Unit tests are replicated for conv tests to exercise per_channel conv. This was not strictly needed since conv kernels were changed such that they did per channel anyway. When per channels is not needed zero point and requant scale were same across channels. 
However for depthwise convolutions we will be using different set of kernels to do per channel, which required us to introduce per_channel member to conv_param structure, to know which kernels to use  for depthwise conv. 
Ensuing modifications were to keep everything in sync for both regular conv and depthwise so that we dont have caveat when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv. Depthwise convs are still pending.
Approach is same as with zero points. Pointer to requantization array is passed and it is looked up using output_channel_index.
All the kernels are appropriately modified except for the depthwise ones.
Tests are altered to generate per channel zero points and requant scales.
Unit tests are replicated for conv tests to exercise per_channel conv. This was not strictly needed since conv kernels were changed such that they did per channel anyway. When per channels is not needed zero point and requant scale were same across channels. 
However for depthwise convolutions we will be using different set of kernels to do per channel, which required us to introduce per_channel member to conv_param structure, to know which kernels to use  for depthwise conv. 
Ensuing modifications were to keep everything in sync for both regular conv and depthwise so that we dont have caveat when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8conv/4x8-aarch32-neon.S:
```diff
@@ -100,14 +108,10 @@ BEGIN_FUNCTION pytorch_q8conv_ukernel_4x8__aarch32_neon
     # - d14 = a_zero_point
     VLD1.8 {d14[]}, [r9]
-    ADD r9, r9, 4
+    # add 8 bytes to get to vfmax
```
Sorry did not fix it here. It gets fixed by the perf related PR.

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv. Depthwise convs are still pending.
Approach is same as with zero points. Pointer to requantization array is passed and it is looked up using output_channel_index.
All the kernels are appropriately modified except for the depthwise ones.
Tests are altered to generate per channel zero points and requant scales.
Unit tests are replicated for conv tests to exercise per_channel conv. This was not strictly needed since conv kernels were changed such that they did per channel anyway. When per channels is not needed zero point and requant scale were same across channels. 
However for depthwise convolutions we will be using different set of kernels to do per channel, which required us to introduce per_channel member to conv_param structure, to know which kernels to use  for depthwise conv. 
Ensuing modifications were to keep everything in sync for both regular conv and depthwise so that we dont have caveat when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv. Depthwise convs are still pending.
Approach is same as with zero points. Pointer to requantization array is passed and it is looked up using output_channel_index.
All the kernels are appropriately modified except for the depthwise ones.
Tests are altered to generate per channel zero points and requant scales.
Unit tests are replicated for conv tests to exercise per_channel conv. This was not strictly needed since conv kernels were changed such that they did per channel anyway. When per channels is not needed zero point and requant scale were same across channels. 
However for depthwise convolutions we will be using different set of kernels to do per channel, which required us to introduce per_channel member to conv_param structure, to know which kernels to use  for depthwise conv. 
Ensuing modifications were to keep everything in sync for both regular conv and depthwise so that we dont have caveat when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv. Depthwise convs are still pending.
Approach is same as with zero points. Pointer to requantization array is passed and it is looked up using output_channel_index.
All the kernels are appropriately modified except for the depthwise ones.
Tests are altered to generate per channel zero points and requant scales.
Unit tests are replicated for conv tests to exercise per_channel conv. This was not strictly needed since conv kernels were changed such that they did per channel anyway. When per channels is not needed zero point and requant scale were same across channels. 
However for depthwise convolutions we will be using different set of kernels to do per channel, which required us to introduce per_channel member to conv_param structure, to know which kernels to use  for depthwise conv. 
Ensuing modifications were to keep everything in sync for both regular conv and depthwise so that we dont have caveat when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to enable per channel requant."


Summary:
Now channel wise quantization is supported for linear/conv. Depthwise convs are still pending.
Approach is same as with zero points. Pointer to requantization array is passed and it is looked up using output_channel_index.
All the kernels are appropriately modified except for the depthwise ones.
Tests are altered to generate per channel zero points and requant scales.
Unit tests are replicated for conv tests to exercise per_channel conv. This was not strictly needed since conv kernels were changed such that they did per channel anyway. When per channels is not needed zero point and requant scale were same across channels. 
However for depthwise convolutions we will be using different set of kernels to do per channel, which required us to introduce per_channel member to conv_param structure, to know which kernels to use  for depthwise conv. 
Ensuing modifications were to keep everything in sync for both regular conv and depthwise so that we dont have caveat when reading the code, that why does depthwise have separate test for
per channel and non-depthwise conv does not.

Test Plan:
Via tests inside qnnpack, i.e., q8gemm-test, q8conv/dwconv test.
fully-conntected-test, convolution-test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21339041](https://our.internmc.facebook.com/intern/diff/D21339041)

[ghstack-poisoned]
